sound/soc/sof/intel/hda-loader.c

```diff
@@ -551,14 +551,15 @@ int hda_dsp_ipc4_load_library(struct snd_sof_dev *sdev,
 			      struct sof_ipc4_fw_library *fw_lib, bool reload)
 {
 	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
 	struct hdac_ext_stream *hext_stream;
 	struct firmware stripped_firmware;
 	struct sof_ipc4_msg msg = {};
 	struct snd_dma_buffer dmab;
 	int ret, ret1;
 
-	/* IMR booting will restore the libraries as well, skip the loading */
-	if (reload && hda->booted_from_imr)
+	/* if IMR booting is enabled and fw context is saved for D3 state, skip the loading */
+	if (reload && hda->booted_from_imr && ipc4_data->fw_context_save)
```

Comment:
it's a bit odd to have different structures to stored 'booted_from_imr' and 'fw_context_save'. Can we put all the information in the same place?

Reply:
@plbossart  booted_from_imr is a intel feature which is set in sof/intel layer. fw_context_save is a common feature which is set in sof core layer. So I don't prefer to combine them in hda structure.